src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+        };
+        try timestamps_array.append(metadata_array.items[0].start_time);
+
+        for (1..metadata_array.items.len) |index| {
+            const current_metadata = metadata_array.items[index];
+
+            if ((current_metadata.lower_bound_slope <= merge_metadata.upper_bound_slope) and
+                (current_metadata.upper_bound_slope >= merge_metadata.lower_bound_slope))
+            {
+                // The current segment metadata can be merged. Update the bounds's slopes.
+                try timestamps_array.append(current_metadata.start_time);
+                merge_metadata.lower_bound_slope = @max(
+                    merge_metadata.lower_bound_slope,
+                    current_metadata.lower_bound_slope,
+                );
+                merge_metadata.upper_bound_slope = @min(
+                    merge_metadata.upper_bound_slope,
+                    current_metadata.upper_bound_slope,
+                );
+            } else {
+                // The current segment metadata cannot be merged. Append the merged metadata and
+                // create a new one.
+                for (timestamps_array.items) |timestamp| {
+                    try merged_segments_metadata.append(.{
+                        .start_time = timestamp,
+                        .interception = merge_metadata.interception,
+                        .lower_bound_slope = merge_metadata.lower_bound_slope,
+                        .upper_bound_slope = merge_metadata.upper_bound_slope,
+                    });
+                }
+                timestamps_array.clearRetainingCapacity();
+                merge_metadata = .{
+                    .start_time = 0.0,
+                    .interception = current_metadata.interception,
+                    .lower_bound_slope = current_metadata.lower_bound_slope,
+                    .upper_bound_slope = current_metadata.upper_bound_slope,
+                };
+                try timestamps_array.append(current_metadata.start_time);
+            }
+        }
+
+        if (timestamps_array.items.len > 0) {
+            // Append the final merged segment metadata.
+            for (timestamps_array.items) |timestamp| {
+                try merged_segments_metadata.append(.{
+                    .start_time = timestamp,
+                    .interception = merge_metadata.interception,
+                    .lower_bound_slope = merge_metadata.lower_bound_slope,
+                    .upper_bound_slope = merge_metadata.upper_bound_slope,
+                });
+            }
+            timestamps_array.clearRetainingCapacity();
+        }
+    }
+
+    // This step is needed since the timestamp order is lost due to the hashmap.
+    mem.sort(
+        SegmentMetadata,
+        merged_segments_metadata.items,
+        {},
+        compareMetadataByStartTime(),
+    );
+}
+
+/// Sim-Piece Phase 3. Compute a hash map from interception points in `merged_segments_metadata`
```

Comment:
I agree, the word "compute" can be unclear here. Another option could be "populate", but I am not sure, and it is also unclear. Seems like any possible option can receive the same comment. Thus, I leave it as it is, but improved the comment as suggested.

Reply:
In this case I would use "populate" in both the comment and in the function name as the function seems to populate `merged_segments_metadata_map` based on the elements in `merged_segments_metadata` and thus I think populate better describes what the function does than compute. However, if the function was calculating a single result from `merged_segments_metadata` I would probably use compute instead of populate.

Reply:
Fixed. Changed to populate.